servicetalk-http-api/src/main/java/io/servicetalk/http/api/ProxyConfigBuilder.java

```diff
@@ -0,0 +1,128 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import java.util.function.Consumer;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Builder for {@link ProxyConfig}.
+ *
+ * @param <A> the type of address
+ */
+public final class ProxyConfigBuilder<A> {
+
+    private static final Consumer<HttpHeaders> NOOP_HEADERS_CONSUMER = new Consumer<HttpHeaders>() {
+        @Override
+        public void accept(final HttpHeaders headers) {
+        }
+
+        @Override
+        public String toString() {
+            return "NOOP";
+        }
+    };
+
+    private final A address;
+    private Consumer<HttpHeaders> connectRequestHeadersInitializer = NOOP_HEADERS_CONSUMER;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param address Proxy address
+     * @see ProxyConfig#address()
+     */
+    public ProxyConfigBuilder(final A address) {
+        this.address = requireNonNull(address);
+    }
+
+    /**
+     * Sets an initializer for {@link HttpHeaders} related to
+     * <a href="https://datatracker.ietf.org/doc/html/rfc9110#section-9.3.6">HTTP/1.1 CONNECT</a> request.
+     *
+     * @param connectRequestHeadersInitializer {@link Consumer} that can be used to set custom {@link HttpHeaders} for
+     * {@code HTTP/1.1 CONNECT} request (auth, tracing, etc.)
+     * @return {@code this}
+     * @see ProxyConfig#connectRequestHeadersInitializer()
+     */
+    public ProxyConfigBuilder<A> connectRequestHeadersInitializer(
+            final Consumer<HttpHeaders> connectRequestHeadersInitializer) {
+        this.connectRequestHeadersInitializer = requireNonNull(connectRequestHeadersInitializer);
+        return this;
+    }
+
+    /**
+     * Builds a new {@link ProxyConfig}.
+     *
+     * @return a new {@link ProxyConfig}.
+     */
+    public ProxyConfig<A> build() {
+        return new DefaultProxyConfig<>(address, connectRequestHeadersInitializer);
+    }
+
+    private static final class DefaultProxyConfig<A> implements ProxyConfig<A> {
+
+        private final A address;
+        private final Consumer<HttpHeaders> connectRequestHeadersInitializer;
+
+        private DefaultProxyConfig(final A address, final Consumer<HttpHeaders> connectRequestHeadersInitializer) {
+            this.address = address;
+            this.connectRequestHeadersInitializer = connectRequestHeadersInitializer;
+        }
+
+        @Override
+        public A address() {
+            return address;
+        }
+
+        @Override
+        public Consumer<HttpHeaders> connectRequestHeadersInitializer() {
+            return connectRequestHeadersInitializer;
+        }
+
+        @Override
+        public boolean equals(final Object o) {
```

Comment:
Because one of the fields is a `Consumer`, which is going to often be set by anonymous lambda, I think notions of equality and hash-code are going to be error prone. Does this class need to have such notions?

Reply:
I think it's fine to keep best effort and we can rely on users to care about equality for consumer. If they don't have, having different lambdas will make configs different. if they care, they must provide the same instance of Consumer or override equals/hashCode for consumer.

Reply:
If that's an invariant which is important for the semantics of the executed code, I think we should point this out in a javadoc somewhere.

Reply:
It feels kind of obvious that if users pass 2 different lambdas, the resulting "pojo" will be different. Note that there is no use-case for equals/hashCode right now to really worry about it. This is a pkg-private class, which makes it harder to document impl details. Not sure if we need to put this info on interface bcz it's not required right now.